core/arch/arm/kernel/user_ta.c

```diff
@@ -687,6 +689,35 @@ static TEE_Result add_deps(struct user_ta_ctx *utc __unused,
 
 #endif
 
+#ifdef CFG_TA_ASLR
+static size_t aslr_offset(size_t min, size_t max)
+{
+	uint32_t rnd32 = 0;
+	size_t rnd = 0;
+
+	if (crypto_rng_read(&rnd32, sizeof(rnd)))
+		return 0;
+	if (max > min)
+		rnd = rnd32 % (max - min);
+	return (min + rnd) * CORE_MMU_USER_CODE_SIZE;
+}
+#else
+static size_t aslr_offset(size_t min, size_t max)
+{
+	return 0;
+}
+#endif
+
+static vaddr_t get_stack_va_hint(void)
+{
+	vaddr_t base = 0;
+
+	core_mmu_get_user_va_range(&base, NULL);
+
+	/* 1 spare page for the ---R-X page + 1 guard page */
```

Comment:
Why is the guard page needed?

Reply:
I think it's because of this https://github.com/OP-TEE/optee_os/blob/a67a20cb7f93/core/arch/arm/mm/tee_mmu.c#L61 but I must double check